<launch>

  <arg name="ouster_ns" default="ouster" doc="Override the default namespace of all ouster nodes"/>
  <arg name="sensor_hostname" default="" doc="hostname or IP in dotted decimal form of the sensor"/>
  <arg name="udp_dest" default="" doc="hostname or IP where the sensor will send data packets"/>
  <arg name="lidar_port" default="0" doc="port to which the sensor should send lidar data"/>
  <arg name="imu_port" default="0" doc="port to which the sensor should send imu data"/>
  <arg name="udp_profile_lidar" default="" doc="lidar packet profile; possible values: {
    LEGACY,
    RNG19_RFL8_SIG16_NIR16_DUAL,
    RNG19_RFL8_SIG16_NIR16,
    RNG15_RFL8_NIR8
    }"/>
  <arg name="lidar_mode" default="" doc="resolution and rate; possible vaules: {
    512x10,
    512x20,
    1024x10,
    1024x20,
    2048x10
    }"/>
  <arg name="timestamp_mode" default="" doc="method used to timestamp measurements; possible values: {
    TIME_FROM_INTERNAL_OSC,
    TIME_FROM_SYNC_PULSE_IN,
    TIME_FROM_PTP_1588,
    TIME_FROM_ROS_TIME
    }"/>
  <arg name="metadata" default="" doc="path to write metadata file when receiving sensor data"/>
  <arg name="bag_file" default="" doc="file name to use for the recorded bag file"/>
  <arg name="viz" default="true" doc="whether to run a rviz"/>
  <arg name="rviz_config" default="$(find ouster_ros)/config/viz.rviz" doc="optional rviz config file"/>
  <arg name="tf_prefix" default="" doc="namespace for tf transforms"/>

  <group ns="$(arg ouster_ns)">
    <node pkg="nodelet" type="nodelet" name="os_nodelet_mgr"
      output="screen" required="true"
      launch-prefix="bash -c 'sleep 2; $0 $@' "
      args="manager"/>
  </group>

  <group ns="$(arg ouster_ns)">
    <node pkg="nodelet" type="nodelet" name="os_node"
      output="screen" required="true"
      launch-prefix="bash -c 'sleep 3; $0 $@' "
      args="load nodelets_os/OusterSensor os_nodelet_mgr">
      <param name="~/sensor_hostname" type="str" value="$(arg sensor_hostname)"/>
      <param name="~/udp_dest" type="str" value="$(arg udp_dest)"/>
      <param name="~/lidar_port" type="int" value="$(arg lidar_port)"/>
      <param name="~/imu_port" type="int" value="$(arg imu_port)"/>
      <param name="~/udp_profile_lidar" type="str" value="$(arg udp_profile_lidar)"/>
      <param name="~/lidar_mode" type="str" value="$(arg lidar_mode)"/>
      <param name="~/timestamp_mode" type="str" value="$(arg timestamp_mode)"/>
      <param name="~/metadata" type="str" value="$(arg metadata)"/>
    </node>
  </group>

  <include file="$(find ouster_ros)/launch/common.launch">
    <arg name="ouster_ns" value="$(arg ouster_ns)"/>
    <arg name="viz" value="$(arg viz)"/>
    <arg name="rviz_config" value="$(arg rviz_config)"/>
    <arg name="tf_prefix" value="$(arg tf_prefix)"/>
    <arg name="timestamp_mode" value="$(arg timestamp_mode)"/>
  </include>

  <arg name="_use_bag_file_name" value="$(eval not (bag_file == ''))"/>
  <arg name="_topics_to_record" value="/$(arg ouster_ns)/imu_packets /$(arg ouster_ns)/lidar_packets"/>

  <node if="$(arg _use_bag_file_name)" pkg="rosbag" type="record" name="rosbag_record_sensor"
       output="screen" required="true"
       args="record -O $(arg bag_file) $(arg _topics_to_record)"/>

  <node unless="$(arg _use_bag_file_name)" pkg="rosbag" type="record" name="rosbag_record_sensor"
       output="screen" required="true"
       args="record $(arg _topics_to_record)"/>

</launch>
